Refactor/consolidate net pkt ieee802154 options

fgrandel opened the pull request:
This PR addresses part 2 of the initial development approach outlined in https://github.com/zephyrproject-rtos/zephyr/issues/50336#issuecomment-1250250154 and confirmed in https://github.com/zephyrproject-rtos/zephyr/issues/50336#issuecomment-1250831121:

> Consolidate IEEE802.15.4 options in net_pkt

The detailed approach and rationale for this change is outlined in the linked RFC.

The PR also fixes several bugs found while working on this change. Bugs are fixed in separate commits and have been reported as separate issues so they can be backported if you wish so:

Fixes: #51261
Fixes: #51263 
Fixes: #51264 
Fixes: #51265

Commit:
drivers: ieee802154: don't allocate rx pkt from tx pool

Several IEEE 802154 drivers allocated RX packets from the TX pool.

This may seem like a minor problem at first sight but it may become
problematic if the pool is used to distinguish package types as is the
case in some code paths, e.g. for packet priority or determination of
the packet buffer pool.

This bug also has the potential of starving the TX pool capacity which
even may make devices vulnerable to DoS attacks as sending may be
prohibited by addressing enough RX packets to a device to let it run out
of TX capacity.

Fixes: #51261

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: fix minor typos

This change just fixes some insignificant typos.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
drivers: ieee802154: nrf: wrapped pkt attribute access

Nordic's IEEE 802.15.4 radio driver adapter layer had a few raw accesses
to net packet attributes.

Packet attributes should never be accessed directly, though, but only
through the dedicated accessor methods provided by the net core.

This change replaces raw accesses to packet attributes by their
respective wrapper functions.

This also is a necessary precondition to the isolation and
encapsulation of IEEE 802.15.4 specific packet attributes which will be
introduced in a later commit of this change set.

Fixes: #51264

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: ip: fix and consolidate cloning of ll addr

The net packet structure contains pointers to link-layer source and
destination addresses. Usually, these structures do not point to
separately allocated memory but directly into the packet's data buffer.

In case of a deep package clone (which includes copying the buffer) the
copy of the ll addresses continued to point into the old package
(contrary to a rather misleading inline comment). This was proven by an
additional failing unit test assertion.

As the original package may be unreferenced while the cloned package is
still being accessed, the ll address pointers of the cloned package may
become invalid.

The fix consists of two parts:
 * First it is determined whether a given ll address actually points into
   the buffer and if so at which logical cursor offset it is located.
 * If the address points into the package buffer then the cursor API is
   used to determine the corresponding physical memory location in the
   cloned package. The ll address of the cloned package is then patched
   to point to the cloned buffer.

Additional assertions were introduced to the existing unit test to ensure
that the newly generated address points to the correct content of the
cloned package.

The solution is implemented in a generic way so that the previously
redundant implementations were consolidated into a single one. The code
includes a check that ensures that the ll address check and manipulation
will be skipped in case of shallow package copies.

The change also addresses problems related to the "overwrite" flag of the
package:
 * Package cloning assumes the overwrite flag to be set. Otherwise it
   will not work correctly. This was not ensured inside the clone method.
 * Package cloning manipulates the overwrite flag of the cloned package
   but does not reset it to represent the same state as the original
   package.

The change introduces a fix and unit test assertions for both problems.

Fixes: #51265

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: set ll protocol in pkt

The IEEE 802.15.4 L2 now sets the ll protocol in the packet to a
specific value. This corresponds to the respective solution in Linux and
is required to validate access to IEEE 802.15.4 specific attributes of
the packet.

Later change sets will rely on this value to ensure that IEEE 802.15.4
specific package content can only be accessed on IEEE 802.15.4 packages.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: pkt: decouple from ieee802154 internals

This change implements part two of the program laid out in the TSCH RFC,
see #50336#issuecomment-1250250154 :

> Consolidate IEEE 802.15.4 options in net_pkt

This change improves decoupling of generic net core code from
IEEE 802.15.4 internals. It also simplifies IEEE 802.15.4
attribute cloning and thereby makes it easier to maintain and less
error prone (and probably even faster as individual bits are no longer
copied over separately).

This enables us to extend and design IEEE 802.15.4 L2 attributes inside
the package in isolation from the net core code which will no longer
have to be changed when introducing changes or additions to the flags.

This flexibility will be built upon in later change sets to model the
IEEE 802.15.4 attributes closer to the spec.

The solution is inspired by Linux's sk_buff->cb attribute which addresses
the same concern as the attribute introduced in this change set:
https://elixir.bootlin.com/linux/v6.0.1/source/include/linux/skbuff.h#L871

As the inline comment says: The cb attribute can be made a union or even a
uint8[something] in the future, if further L2s need a control block, too.
Right now such full indirection would make the code overly abstract, so
I chose to compromise with maintainability in mind.

Care has been taken to ensure that this changes does not introduce
additional padding into the net package. To maintain zero-padding, future
changes to the net packet struct will have to ensure that the
IEEE 802.15.4 struct is 4-byte aligned (iff the IEEE 802.15.4 struct
continues with max uint32_t scalar members) which is no deviation from
the previous implementation.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: pkt: optimize net_pkt packing

This change removes unnecessary padding from the net_pkt struct in
preparation for restructuring it in a later commit. Depending on the
number of activated Kconfig options, this saves between 4 and 8 bytes
per packet.

The change also removes anonymous unions inside of bitfields:
 * The assumption, that anonymous unions inside bitfields save space is
   wrong. They rather enforce an additional byte aligment boundary with
   padding before or after on most compilers (try -Wpadded to prove it).
   Removing the unions therefore counter-intuitively removes padding.
 * Some of the fields inside the union may actually not be orthogonal if
   several extensions are enabled at the same time. They may overwrite
   each other when activated at the same time.

The change therefore not only improves packing of net_pkt but also makes
net_pkt easier to understand and maintain by removing potentially
non-orthogonal uses of bits within bitfields.

The union of RX/TX-only attributes has been maintained, though, as it DOES
save space and is easy to maintain. Maintainability has further been
improved by introducing additional inline comments and anonymous structs
that make the orthogonality of RX- and TX-attributes even more obvious.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

nandojve commented on subsys/net/l2/ieee802154/ieee802154.c:
```diff
@@ -237,6 +238,8 @@ static enum net_verdict ieee802154_recv(struct net_if *iface, struct net_pkt *pk
 
 	ieee802154_acknowledge(iface, &mpdu);
 
+	net_pkt_set_ll_proto_type(pkt, ETH_P_IEEE802154);
+
```
I'm not feel comfortable to have a dependency from ethernet inside IEEE 802.15.4.

The set_pkt_ll_addr do this after all:

https://github.com/zephyrproject-rtos/zephyr/blob/2037b20db52a40d8639ea25647b362e4a4fa15e4/subsys/net/l2/ieee802154/ieee802154.c#L100

Is not possible to consult net_linkaddr->addr->type later ?

fgrandel replied:
@nandojve I'm afraid there's no short answer to this... I did quite a lot of research before choosing this place as I was equally surprised when first looking at the Linux kernel and socket UAPI. But as a matter of fact `linux/if_ether.h` provides the (socket) PHY layer categorization for many network stacks - not just ethernet (albeit not for all - e.g. bluetooth seems to be an exception). See https://elixir.bootlin.com/linux/v6.0.1/A/ident/ETH_P_IEEE802154 for IEEE 802.15.4 or https://elixir.bootlin.com/linux/v6.0.1/C/ident/ETH_P_CAN for another PHY relevant to Zephyr. These constants even made it into userspace (see e.g. https://github.com/linux-wpan/wpan-tools/blob/master/examples/af_packet_rx.c for IEEE 802.15.4 or more generally `man 7 packet`, `sll_protocol`).

Zephyr has chosen to model their socket API following Linux conventions (which is not a bad choice IMHO). As we provide IEEE 802.15.4 RX socket support in `ieee802154_recv()` in `ieee802154.c`, this is IMO the correct place to set the protocol type as it is consistent with the Linux implementation and Zephyr ethernet stack's `ether_recv()`.

As an aside: The fact that Zephyr's CAN L2 does not set this attribute although it has its own socket stack is IMHO an oversight (as it was prior to this change in IEEE 802.15.4). Bluetooth does not yet support low level sockets in Zephyr, so maybe that's why they don't set a protocol, although Linux's bluetooth implementation does so?

My understanding is that `net_linkaddr.type` is something different. It is not generally true that there is a single protocol for a given link layer address nor the contrary. So `net_linkaddr.type` and `net_pkt.ll_proto_type` are orthogonal IMO.

The doxygen documentation is at best ambiguous and at worst plain wrong calling `net_linkaddr` a "hardware" address rather than a link-layer address. There is no global structure like `net_linkaddr` in Linux AFAICS. The details of LL addresses are usually kept privately in `skbuff->cb` or in opaque "polymorphic" protocol family specific socket addresses which expose the address family of the address so that the address can be cast to the protocol specific implementation. I'm not sure why `net_linkaddr.type` has its own global configuration space rather than referring to AF_.../PF_ constants. Is it maybe just legacy of pre-socket Zephyr code?

I'd therefore argue the other way round: `net_linkaddr.type`/`enum net_link_type` is redundant and should probably be replaced by polymorphic socket addresses whose internal polymorphic structure depends on the publicly visible AF_... constant. These structures already exist in Zephyr and could easily be used to replace `net_linkaddr` AFAICS.

The only uses of `net_linkaddr.type` I can find in the code are to identify 6lo headers and generate IPv6 addresses out of LL addresses. Both would be better served by opaque socket addresses and delegated to L2 rather than kept in net core breaking encapsulation as is now the case. But all this is definitely out-of-scope for the current change set.

nandojve replied:
It seems there is not much space to set that info. I'm thinning that maybe a solution inside **net_pkt** could avoid to contaminate any L2 with Ethernet stuff. Do you think it is valid introduce something like below at **net_pkt**?

```diff
+void net_pkt_set_ll_proto_type_<by_ll_type>(struct net_pkt *pkt, enum net_link_type type) {
+	switch (type) {
+	...
+	case NET_LINK_IEEE802154:
+		net_pkt_set_ll_proto_type(pkt, ETH_P_IEEE802154);
+		break;
+	...
+	}
+}
```

fgrandel replied:
IMO this is much worse as it introduces more dependencies on specific L2 stacks into net core while the solution I propose does not break encapsulation anywhere plus it is in line with what ethernet.h obviously claims to be (i.e. a counterpart to if_ether.h). Let's see what others think about this topic. Of course I'll be convinced if your proposal becomes the majority vote.

rlubos replied:
Personally, I'd stick to the current proposal.

@nandojve While the solution you propose may seem simple for IEEE802154, things are getting more complicated for Ethernet, where the actual proto type is not fixed, but rather parsed from the L2 header (EtherType field). And it gets even more complex with VLAN involved. All this parsing logic would need to be covered by this generic function, which is not the best choice IMHO.

nandojve reviewed (changes requested):
Hi @fgrandel ,

I looked inside 802.14.2 only and have doubts about 2 points.

fgrandel commented on drivers/ieee802154/ieee802154_b91.c:
```diff
@@ -394,7 +394,7 @@ static void b91_iface_init(struct net_if *iface)
 	b91->iface = iface;
 
-	ieee802154_init(iface);
+	ieee802154_init(iface, 0);
```
You're right of course. Must be fixed before merging.

fgrandel replied:
Fixed.

Commit:
drivers: ieee802154: properly announce promisc mode

Most IEEE 802.15.4 drivers do not support promiscuous mode, some do.
There is a dedicated L2 flag to signal this capability to clients.

Unfortunately the IEEE 802.15.4 L2 stack does not announce this flag
even for drivers that correctly expose it in their HW capabilities.
Some clients (notably the OpenThread L2) even uses promiscuous mode
without checking whether the driver actually supports it.

This change lets the vanilla IEEE 802.15.4 L2 check the driver's
HW capabilities to announce promiscuous mode on its 'get_flags()'
interface if supported.

The OpenThread L2 uses a constant (potentially incorrect) response
to 'get_flags()'. Fixing the OpenThread L2 is out of scope of this
change. This change just introduces TODO messages to the OpenThread code
so that the OpenThread team may fix the issue (or delete the TODO if they
deem it irrelevant).

Fixes: #51263

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

fgrandel commented:
@nandojve Thanks for your review! Could you have a look at my fix and answer to your comment to see whether this makes sense now?

rlubos reviewed (commented):
Looks good, just some nits.

fgrandel commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -530,6 +530,11 @@ static int openthread_init(struct net_if *iface)
 
 void ieee802154_init(struct net_if *iface)
 {
+	/* TODO: openthread_flags() should report NET_L2_PROMISC_MODE
+	 *       if the radio driver reports the IEEE802154_HW_PROMISC
+	 *       capability.
+	 */
+
```
True. Fixed.

fgrandel commented on include/zephyr/net/ieee802154_pkt.h:
```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
+ *
```
Ok, I added a second line with the current year and my name. I guess the original copyright should not be changed for legal reasons. I'd also feel bad to remove it completely as most of the code was just copied over.

rlubos replied:
Yes, that's what I had in mind - extend the list, not replace the existing copyright.

fgrandel commented:
@rlubos Fixed the things you remarked. I'm a bit unsure about the copyright stuff. Is that what you meant?

rlubos reviewed (approved):
LGTM, thanks
